ledger-service/http-json/src/main/scala/com/digitalasset/http/Endpoints.scala

```diff
@@ -77,17 +79,84 @@ class Endpoints(
   private def toRoute(res: => Future[Error \/ SearchResult[Error \/ JsValue]]): Route =
     responseToRoute(httpResponse(res))
 
+  // Always put this directive after a path to ensure
+  // that you don't log request bodies multiple times (simply because a matching test was made multiple times).
+  // TL;DR JUST PUT THIS THING AFTER YOUR FINAL PATH MATCHING
+  def logJsonRequestAndResult(implicit
+      lc: LoggingContextOf[InstanceUUID with RequestID]
+  ): Directive0 =
+    extractRequest & extractClientIP tflatMap { case (request, remoteAddress) =>
+      def logWithHttpMessageBodyIfAvailable(
+          httpMessage: HttpMessage,
+          msg: String,
+          bodyKind: String,
+      ): Future[Unit] =
+        if (
+          httpMessage
+            .header[`Content-Type`]
+            .map(_.contentType)
+            .contains(ContentTypes.`application/json`)
+        )
+          httpMessage
+            .entity()
+            .toStrict(maxTimeToCollectRequest)
+            .map(_.data.utf8String)
+            .map(_.parseJson)
+            .map(body =>
+              withEnrichedLoggingContext(
+                LoggingContextOf.label[RequestEntity],
+                s"$bodyKind-body" -> body,
+              )
+                .run(implicit lc => logger.debug(msg))
+            )
+            .recover { case ex =>
+              logger.error("Failed to extract body for logging", ex)
+            }
+        else Future.successful(logger.debug(msg))
+      mapRouteResultFuture { responseF =>
+        for {
+          _ <- logWithHttpMessageBodyIfAvailable(
+            request,
+            s"Incoming ${request.method} request on ${request.uri} from $remoteAddress",
+            "request",
+          )
+          response <- responseF
+          _ <- response match {
+            case Complete(httpResponse) =>
+              logWithHttpMessageBodyIfAvailable(
+                httpResponse,
+                s"Responding to client with HTTP ${httpResponse.status}",
+                "response",
+              )
+            case _ =>
+              Future.failed _ apply
+                new RuntimeException(
+                  """Logging the request & response should never happen on routes which get rejected.
+                    |Make sure to place the directive only at places where a match is guaranteed (e.g. after the path directive).""".stripMargin
+                )
+          }
+        } yield response
+      }
+    }
```

Comment:
This whole body logging should be fully plugable, because in a production scenario you don't want to log the incoming request and outgoing response bodies (as these are potentially big and let the logs explode with time next to the performance implications).